### ISSUE TITLE
[Feature] Running the Cap'n inside a docker container

### ISSUE THREAD

**Issue by shochdoerfer** [CaptainHookPhp/captainhook#29]
After having mentioned the Cap'n in one of my talks, I was asked how to run git hooks when the application is running in a docker container and no php cli is available on the host. I think that is an interesting use-case, especially when you want to offer hooks for other frameworks & tools as many people these days have their own docker setup in place.

Letting the hooks run commands in a container could be a trivial thing, instead of the php based hooks, provide a shell script that would call the php hooks in docker. The container name to run would need to configurable, I guess.

However, the problem is, how to install the Cap'n without invoking Composer/Php on the host. You could potentially provide a 2nd way of installing the Cap'n via some Bash tooling then. Not sure though if it really is feasible to maintain 2 separate installers.

**Comment by sebastianfeldmann** [comment]
I'm far from a Docker pro so this is was I understand has to happen.
- You need a Docker container able to run `php cli` and `git`
- In case you have no php to execute locally you have to start a docker container able to execute `CaptainHook` with the php cli and has access to the source code and the repository so it can execute some git commands as well.
- Create the real git hooks inside the .git directory without using php

I could think of a `captainhook/docker` packages that could provide an example docker file and some scripts to install the `docker hooks` to the local directory.

The `docker hooks` would have to:
- start the container
- execute captainhook somehow and pass the original arguments in a way they can be handled inside the container. I think especially tricky are the commit-message arguments regarding the file location of the current commit message file, so it can be accessed inside the container
- retrieve the result from inside the container and exit with 0 or 1 depending on the received result.

Is this about right?

**Comment by shochdoerfer** [comment]
That's correct. You can invoke commands in docker either via docker cli or via docker-compose like this: `docker-compose exec the-container-name /var/www/html/vendor/bin/captainhook`

I thought it might be tricky to pass parameters of the hook itself into the container, I guess for most parameters things should be simple, e.g. commit Id. If in case of the commit-message hook you need to pass an absolute path, one might need to be able to configure some path mapping then.

**Comment by sebastianfeldmann** [comment]
If we put this in a separate package we can define another binary script like `captainhook.docker.bridge` that gets called by the docker cli and does the following things:
- mapping the arguments to the captain hook commands
- mappging path arguments to a path that works inside the container
- calling the right original captain hook commands
- returning the errors and exit codes in a way the docker cli can communicate it to the originally called hook script

This can still be done on PHP since it runs inside the container. The only stuff where we have to leave php behind is for the part that is called by git and triggers the docker cli.
And the hook installation of course.

Sounds doable I would say.

**Comment by dkreuer** [comment]
Hook installation can still be done with php when done with the composer plugin. Maybe a golang wrapper to call forth the captain is possible? It could detect the environment, required tools etc. But would be a major project itself.

**Comment by icanhazstring** [comment]
I think the only thing that has to be changed, is the git hook content itself.
Currently CaptainHook writes PHP code inside the `.git/hooks/commit-msg` (for example).

What needs to be changed to run this into docker (or any other virtual machine) would be to postpone the php execution. What we need to do?

1. CaptainHook does *NOT* install the php code into the local `.git/hooks` itself but rather provides a single php file for execution. This could be as well reside beside the hooks itself - lets say `captain-runner.php`
2. The `hooks` itself might contain something like this:
    ```bash
    #!/usr/bin/env bash
    "`dirname $0`/captain-runner.php" commit-msg "$@"
    ```

   > For this to work for windows, users need to run some kind of cygwin sh on their machine (git for windows)

    As you can see we pass the type of hook into the `captain-runner.php` to we define the run that is running. So Currently the https://github.com/CaptainHookPhp/captainhook/blob/master/src/Hook/Template.php#L52 could be changed to simply tage the first argument of the execution. (Simply a guess, which needs to be checked if its working)
3. Using the `install` command for the captain, we should also ask if the user is running on a virtual machine. If so the hooks in `.git/hooks` have to look a bit different. These must the contain the execution for the docker container as @shochdoerfer pointed out.

    So for example we need to ask for the `container-name` and the `linked-path` where the source resides inside the container itself.

    ```bash
    #!/usr/bin/env bash
    docker-compose exec the-container-name /var/www/html/.git/hooks/captain-runner.php commit-msg "$@"
    ``` 

For this to work tho. I think we should break down this "issue" into multiple ones, or else the impact would be pretty high and the change set is not very good to review.

**Comment by icanhazstring** [comment]
As I talked to @sebastianfeldmann I will provide a PR with a POC. Then we can talk about the final implementation of this. 🙂

**Comment by icanhazstring** [comment]
I have opened up the proper PR.
Feel free to add your thoughts 😃

**Comment by sebastianfeldmann** [comment]
Fixed and released with 4.3.0
